projects/plugins/jetpack/tests/e2e/specs/connection.test.js

```diff
@@ -15,33 +20,49 @@ process.env.SKIP_CONNECT = true;
  * @group connection
  */
 describe( 'Connection', () => {
-	beforeAll( async () => {
-		await execMultipleWpCommands(
-			'wp option delete e2e_jetpack_plan_data',
-			'wp option delete jetpack_active_plan',
-			'wp option delete jetpack_private_options',
-			'wp option delete jetpack_sync_error_idc'
-		);
-		await page.reload();
-		await page.reload();
+	afterEach( async () => {
+		await execShellCommand( 'yarn tunnel-reset' );
```

Comment:
I'm not very fond of the idea of resetting the tunnel during tests. In my mind, these should be totally decoupled and should not know the other one exist. They are different layers, specs being the highest layer, that should not have access to lower layers. The tests should run against a given URL, they should not be able to change that URL.

I ran the tests multiple without resetting the tunnel, and they passed, apparently cleaning on local side only with `resetWordpressInstall` is enough. Maybe we're ok to leave it like that?
Even if they passed, at least 2 of them are actually testing a reconnection, not a fresh connection. Can we live with this? Or do we want to enforce fresh connection?

Reply:
I agree with the idea of decoupling. But I think the only benefit of decoupling is the ability to run _all_ tests in parallel. To do that, we should use a dedicated WP and tunnel for every test. 

Until we'll decide to commit fully in that direction, it feels more like an overengineering.

> at least 2 of them are actually testing a reconnection
Personally, as a QA, I think we should try to aim to test a fresh connection every time. But maybe it is not as important from a technical/business point of view.

Reply:
> To do that, we should use a dedicated WP and tunnel for every test.

I agree. But I think we should do that in the test execution setup, not from the test itself. The execution setup you can control.
If I want to run all the tests against a test site that I have prepared, in the middle of the execution the target url changes and the rest of the tests will run on a different site. This is unexpected and something I would rather avoid in the detriment of convenience. Convenience meaning the ability to run all tests in a single command.

> Personally, as a QA, I think we should try to aim to test a fresh connection every time.

I would aim for a fresh connection every time too. And I'm wondering if there's another way to "clean" a connection. What do we need to clean after the first connection so the next one is considered a fresh one, not a reconnection? I don't know, maybe it's a big effort or maybe it's not possible. But it would be my first try.

So, in the end is a matter of choice in what we're compromising:
* architectural principles (change url mid test)
* test relevance (testing reconnections instead of fresh connections)
* convenience (re-structure tests, have a single test per suite, run suites in parallel, each with their own tunnel)

I honestly don't have my answer to this, neither one of the above sound good to me. And I will be fine with moving on with this approach, but only as a temp solution, with a commitment that we'll revisit it.

Reply:
> But I think we should do that in the test execution setup, not from the test itself

What is a test execution setup? To make tests truly independent so we can run them in parallel, we probably should move all the setup steps into the Jest context. We probably can parallelize in some other ways, but I think Jest's approach is most straightforward. 

> If I want to run all the tests against a test site that I have prepared

Is it a real use case? Why one would want to do that? Our tests not designed in a way to allow running them against any arbitrary URL. And probably, it would not be possible to that at all. 

Our tests are not truly black-boxed, since they rely on low-level site manipulation and mocks, so I don't think we should treat them like that anyway.

Reply:
I think we can agree that the complex nature of Jetpack setup makes it impossible to put all tests in the same "bucket", run them and expect everything to be ok. This even without considering parallelism.

So, we need different buckets (hence the recent `groups` implementation). My rule of thumb for these bucket is: can I run all the tests at once against the same site? If yes, usually they can also run in parallel inside this bucket. That a different topic.

And you control everything. Start an env, Create a tunnel, I run the test(s). This sequence applies for each bucket.

In our case, we would have 3 different buckets, because we want a fresh connection for each. That means 3 suites. In CI, 3 different groups. 2 more than we already have. This is the inconvenience. But for me at least, it brings more clarity. 

> Is it a real use case? Why one would want to do that? Our tests not designed in a way to allow running them against any arbitrary URL. And probably, it would not be possible to that at all.

For me it is. Ok, not all tests, we can exclude the connection ones. 
I know they are not designed for that. But maybe we can start designing them that way? 
I think this is a longer discussion about direction and strategy that we can move to other channels. 

Let's ship it like this, and we'll revisit it later.